### PR TITLE
Various improvements for pycolmap bindings

### PR DESCRIPTION
**Pull request by ahojnnes**
* Resolves some TODOs by adding missing methods, etc.
* Moves the implementation of various `__repr__` to the C++ side.
* Makes Correspondence a data class to benefit from automatic `__copy__` method creation.
* Makes Database object compatible with context manager interface to enable `with pycolmap.Database("path.db") as db` usage. 
* Some backwards compatibility breakage to achieve consistent method names in C++ and Python.

### CODE REVIEW

File: src/colmap/sfm/incremental_triangulator.cc [hunk 1]
```diff
@@ -753,4 +753,12 @@ bool IncrementalTriangulator::HasCameraBogusParams(const Options& options,
   }
 }
 
+std::ostream& operator<<(std::ostream& stream,
+                         const IncrementalTriangulator& triangulator) {
+  stream << "IncrementalTriangulator(reconstruction="
+         << triangulator.reconstruction_ << ", correspondence_graph="
+         << *THROW_CHECK_NOTNULL(triangulator.correspondence_graph_) << ")";
```
Comment (ahojnnes) [inline src/colmap/sfm/incremental_triangulator.cc:760]: Make robust to missing correspondence graph.

Comment (ahojnnes) [reply src/colmap/sfm/incremental_triangulator.cc:760]: Add test.

File: src/colmap/scene/database.h [hunk 2]
```diff
@@ -166,19 +166,19 @@ class Database {
...
   TwoViewGeometry ReadTwoViewGeometry(image_t image_id1,
                                       image_t image_id2) const;
-  void ReadTwoViewGeometries(
-      std::vector<image_pair_t>* image_pair_ids,
-      std::vector<TwoViewGeometry>* two_view_geometries) const;
+  std::vector<std::pair<image_pair_t, TwoViewGeometry>> ReadTwoViewGeometries()
+      const;
 
-  // Read all image pairs that have an entry in the `NumVerifiedImagePairs`
+  // Read all image pairs that have an entry in the `two_view_eg`
```
Comment (B1ueber2y) [inline src/colmap/scene/database.h:178]: what is two_view_eg here?

File: src/colmap/sfm/incremental_triangulator.cc (hunk 1 above)
Comment (B1ueber2y) [reply src/colmap/sfm/incremental_triangulator.cc:760]: Did you add the test somewhere for this? I did not find it.

**Review by B1ueber2y — APPROVED** [review]
Thanks. LGTM

File: src/pycolmap/scene/reconstruction.cc [hunk 3]
```diff
@@ -252,19 +252,16 @@ void BindReconstruction(py::module& m) {
...
-             std::stringstream ss;
-             ss << "Reconstruction(num_reg_images=" << self.NumRegImages()
-                << ", num_cameras=" << self.NumCameras()
-                << ", num_points3D=" << self.NumPoints3D()
-                << ", num_observations=" << self.ComputeNumObservations()
-                << ")";
+             std::ostringstream ss;
+             ss << self;
              return ss.str();
            })
```
Comment (sarlinpe) [inline src/pycolmap/scene/reconstruction.cc:258]: I suggest merging https://github.com/colmap/colmap/pull/2855 first and then changing this to:
`.def("__repr__", &CreateRepresentation<Reconstruction>)`

File: src/pycolmap/sfm/incremental_triangulator.cc [hunk 4]
```diff
@@ -122,7 +122,8 @@ void BindIncrementalTriangulator(py::module& m) {
              return IncrementalTriangulator(self);
            })
       .def("__repr__", [](const IncrementalTriangulator& self) {
```
Comment (sarlinpe) [inline src/pycolmap/sfm/incremental_triangulator.cc:124]: `.def("__repr__", &CreateRepresentation<IncrementalTriangulator>)`

File: src/pycolmap/sfm/observation_manager.cc [hunk 5]
```diff
@@ -134,6 +134,8 @@ void BindObservationManager(py::module& m) {
            "that `IncrementCorrespondenceHasPoint3D` was called for the same"
            "image point and correspondence before.")
       .def("__repr__", [](const ObservationManager& self) {
```
Comment (sarlinpe) [inline src/pycolmap/sfm/observation_manager.cc:136]: `.def("__repr__", &CreateRepresentation<ObservationManager>)`

File: src/pycolmap/scene/correspondence_graph.cc [hunk 6]
```diff
@@ -134,9 +128,8 @@ void BindCorrespondenceGraph(py::module& m) {
              return CorrespondenceGraph(self);
            })
       .def("__repr__", [](const CorrespondenceGraph& self) {
```
Comment (sarlinpe) [inline src/pycolmap/scene/correspondence_graph.cc:130]: `.def("__repr__", &CreateRepresentation<CorrespondenceGraph>)`

**Review by sarlinpe — CHANGES_REQUESTED** [review]

File: src/colmap/sfm/incremental_triangulator.cc (hunk 1 above)
Comment (ahojnnes) [reply src/colmap/sfm/incremental_triangulator.cc:760]: No, this was a note for myself here :-) Done now.

**Review by sarlinpe — APPROVED** [review]
Nice improvements, thanks!

**Pull request merged**